File: src/main/java/com/patres/kotlinplusplus/f_delegate/java/DelegateApp.java
package com.patres.kotlinplusplus.f_delegate.java;

import static java.lang.Thread.sleep;

public class DelegateApp {

    private  Integer price = null;


    private  int calculatePrice() throws InterruptedException {
        if (price == null) {
            sleep(1000);
            price = 10000;
        }
        return price;
    }

    public  void setPrice(Integer price) {
        System.out.println("Old: " + this.price + "| new: " + price);
        this.price = price;
    }
}

File: src/main/java/com/patres/kotlinplusplus/f_delegate/kotlin/DelegateApp.kt
package com.patres.kotlinplusplus.f_delegate.kotlin

import java.lang.Thread.sleep
import kotlin.properties.Delegates.observable

val price: Int by lazy {
    sleep(1000)
    10000
}

var priceObserbavle: Int by observable(0) { property, oldValue, newValue ->
    println("Old: $oldValue | new: $newValue")
}

//var name: String by Delegates.observable("<no name>") { prop, old, new ->
//    println("$old -> $new")
//}

fun main() {
    val sum = 3 * price + 2 * price
    println(sum)
}


private fun calculatePrice(): Int {
    println("Loading...")
    return 10000
}


fun load() = 1